docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,202 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/runtime-client-server.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Client/server implementation of container runtime
+
+## Abstract
+
+A proposal of client/server implementation of container runtime.
+
+## Motivation
+
+Currently, any container runtime has to be linked into the kubelet. This makes experimentation difficult, and prevents users from landing an alternate container runtime without landing code in core kubernetes.
+
+To facilitate experimentation and to enable user choice, this proposal adds a client/server implementation of the container runtime interface.
+
+This implementation will simply encode the requests, send them to a server where they will be decoded and sent into an instance of the container runtime interface.
+
+However, this enables container runtime implementations to be built and maintained outside of the core kubernetes tree.
+
+## Proposed design
+
+**Design of client/server container runtime**
+
+The main idea of client/server container runtime is to keep main control logic in kubelet while letting remote runtime only do dedicated actions:
+
+* A remote runtime manager which implements [Runtime](../../pkg/kubelet/container/runtime.go#L74) interface will be added to kubelet
+* Keep `SyncPod()` in kubelet remote runtime manager and remote runtime only needs to start or kill a pod
+* Keep volume manager in kubelet and pass volumeMap to remote runtime when starting a new pod
+* Keep image garbage collection in kubelet and remote runtime only needs to pull or remove images
+* Merge `RunInContainer` into `ExecInContainer`
+* Avoid `api.Pod` in `RemoteRuntime` interface, but remain essential data for managing pods
```

Comment:
thanks, missed that

On Wed, Dec 9, 2015 at 8:23 PM, feisky <notifications@github.com> wrote:

> In docs/proposals/runtime-client-server.md
> <https://github.com/kubernetes/kubernetes/pull/17048#discussion_r47178620>
> :
>
> > +This implementation will simply encode the requests, send them to a server where they will be decoded and sent into an instance of the container runtime interface.
> > +
> > +However, this enables container runtime implementations to be built and maintained outside of the core kubernetes tree.
> > +
> > +## Proposed design
> > +
> > +**Design of client/server container runtime**
> > +
> > +The main idea of client/server container runtime is to keep main control logic in kubelet while letting remote runtime only do dedicated actions:
> > +
> > +* A remote runtime manager which implements [Runtime](../../pkg/kubelet/container/runtime.go#L74) interface will be added to kubelet
> > +* Keep `SyncPod()` in kubelet remote runtime manager and remote runtime only needs to start or kill a pod
> > +* Keep volume manager in kubelet and pass volumeMap to remote runtime when starting a new pod
> > +* Keep image garbage collection in kubelet and remote runtime only needs to pull or remove images
> > +* Merge `RunInContainer` into `ExecInContainer`
> > +* Avoid `api.Pod` in `RemoteRuntime` interface, but remain essential data for managing pods
>
> Pod is defined at
> https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L155
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17048/files#r47178620>.
>